bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,440 @@
+       */
+      private class CloudBigtableWriter extends Writer<Mutation, Void> {
+        private Connection conn;
+        private BufferedMutator mutator;
+
+        @Override
+        public void open(String uId) throws Exception {
+          conn = new BigtableConnection(config.toHBaseConfig());
+          mutator = conn.getBufferedMutator(TableName.valueOf(config.getTable()));
+        }
+
+        @Override
+        public WriteOperation<Mutation, Void> getWriteOperation() {
+          return CloudBigtableWriteOperation.this;
+        }
+
+        @Override
+        public void write(Mutation value) throws Exception {
+          mutator.mutate(value);
+        }
+
+        @Override
+        public Void close() throws Exception {
+          try {
+            mutator.close();
+          } catch (RetriesExhaustedWithDetailsException e) {
+            List<Throwable> causes = e.getCauses();
+            if (causes.size() == 1) {
+              throw (Exception) causes.get(0);
+            } else {
+              throw e;
+            }
+          }
+          conn.close();
+          return null;
+        }
+      }
+
+      @Override
+      public void initialize(PipelineOptions options) throws Exception {}
+
+      @Override
+      public void finalize(Iterable<Void> writerResults, PipelineOptions options)
+          throws Exception {}
+
+      @Override
+      public Coder<Void> getWriterResultCoder() {
+        return VoidCoder.of();
+      }
+
+      @Override
+      public Sink getSink() {
+        return Sink.this;
+      }
+
+      @Override
+      public Writer<Mutation, Void> createWriter(PipelineOptions options) throws Exception {
+        return new CloudBigtableWriter();
+      }
+    }
+
+    public Sink(CloudBigtableTableConfiguration config) {
+      this.config = config;
+    }
+
+    @Override
+    public void validate(PipelineOptions options) {
+      checkNotNullOrEmpty(config.getProjectId(), "project");
+      checkNotNullOrEmpty(config.getZone(), "zone");
+      checkNotNullOrEmpty(config.getCluster(), "cluster");
+      checkNotNullOrEmpty(config.getTable(), "table");
+      // TODO(sduskis): Validate that the table exists.  This might be tricky since ALPN would have
+      // to be configured on the client side as a bootclasspath.
+    }
+
+    @Override
+    public WriteOperation<Mutation, Void> createWriteOperation(PipelineOptions options) {
+      return new CloudBigtableWriteOperation();
+    }
+  }
+
+  public static com.google.cloud.dataflow.sdk.io.Sink<Mutation> write(
+      CloudBigtableOptions options) {
+    return write(CloudBigtableTableConfiguration.fromCBTOptions(options));
+  }
+
+  public static com.google.cloud.dataflow.sdk.io.Sink<Mutation> write(
+      CloudBigtableTableConfiguration config) {
+    return new Sink(config);
+  }
+
+  public static PTransform<PCollection<Mutation>, PDone> streamingWrite(
```

Comment:
The streamingWrite shouldn't be a separate transform. There should be a single write() that the user provides, that "does the right thing". What is the purpose of distinguishing these? For now, it may be best to have the write PTransform select its behavior from the isStreaming pipeline option, or the bounded/unboundedness of the input PCollection, depending on what the differences in behavior are.